feat(spdx): add SHA-512 hash algorithm support to SPDX serializer

DicksenT opened the pull request:
## Description

This PR adds SHA-512 hash algorithm support to the SPDX SBOM serializer.

Specifically, it updates `spdxChecksums()` in `marshal.go` to handle the `digest.SHA512` case, so Trivy can correctly include SHA-512 checksums in the SPDX output (used in tools like `npm sbom`). This complements #9126, which already added digest and CycloneDX support.

A unit test was also added to ensure SHA-512 is handled correctly in SPDX.

## Related issues
- Closes #9094

## Related PRs
- [x] #9126

## Checklist
- [x] I've read the [guidelines for contributing](https://trivy.dev/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://trivy.dev/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

DicksenT committed:
feat(spdx): add SHA-512 hash algorithm support to SPDX serializer

- Added SHA512 case to marshal.go
- Added new test case for SHA512 digest in marshal_test.go
- Fixed typo in SHA512 algorithm switch

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aquasecurity/trivy?pullRequest=9130) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/trivy?pullRequest=9130) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/trivy?pullRequest=9130) it.</sub>

knqyf263 commented on pkg/sbom/spdx/marshal_test.go:
```diff
@@ -37,6 +38,38 @@ func annotation(t *testing.T, comment string) spdx.Annotation {
 	}
 }
 
+func Test_spdxChecksums_SHA512(t *testing.T) {
```
This test doesn't appear to make sense. You can refer to other tests.

DicksenT replied:
you are right, i will remove the additional function

knqyf263 reviewed (commented):
I think we also need to update unmarshaling. https://github.com/aquasecurity/trivy/pull/9126 will probably help your understanding.

DicksenT commented:
Hi @knqyf263 ,

I’ve addressed the following:

Fixed the digest.SHA512 typo (was mistakenly written as spdx.SHA512)

Removed the incorrect test case and added the correct one for SHA512 support in SPDX

however, I looked into unmarshal.go as suggested, but I couldn’t find explicit handling for SHA1 or SHA256 there.
Should I add new logic specifically for SHA512, or is the current unmarshal logic expected to cover it already?

DicksenT commented:
on it

DicksenT committed:
fix(spdx): resolve lint error in marshal_test.go

DicksenT commented:
@knqyf263  I've fixed the lint issues. Let me know if there's anything else I should address. Thanks again!

knqyf263 commented:
Do you mind signing our CLA?
https://github.com/aquasecurity/trivy/pull/9130#issuecomment-3031841886

DicksenT commented on pkg/sbom/spdx/marshal_test.go:
```diff
@@ -1448,6 +1448,125 @@ func TestMarshaler_Marshal(t *testing.T) {
 			},
 		},
+		{
+			name: "happy path for SHA512 digest",
```
i think it's possible, but maybe i need to tweak marshal.go abit(?), or should i create the test in testdata/happy just like how it done in cycloneDx implementation?

knqyf263 replied:
>but maybe i need to tweak marshal.go a bit(?)

I don't think so. You can just modify the existing case. For example:
https://github.com/aquasecurity/trivy/blob/b42e7fc02116e65077c16812f03baa371280dd2e/pkg/sbom/spdx/marshal_test.go#L441

DicksenT replied:
the test fails with a diff that seems related to Relationship ordering or structure. Would it help if I push the current PR so you can review or debug it directly?

DicksenT replied:
Just pushed the latest changes,  let me know if you need anything else on my end

knqyf263 replied:
I don't have time to debug it right now.

DicksenT replied:
understandable, should we keep using spesific sha512 test case or you want me keep trying to merge the testcase?

DmitriyLewen replied:
simplified the test example for sha512 -  https://github.com/aquasecurity/trivy/pull/9130/commits/6ee1bc953d10c0259b0ee3fbf3af814e35bc8b87

DicksenT committed:
debug: merge testcase for SHA-512 SBOM coverage

DicksenT committed:
debug: merge testcase for SHA-512 SBOM coverage

DmitriyLewen committed:
test: refactor testdata with sha512

DmitriyLewen commented:
Hello @DicksenT 
you also need update `unmarshal` logic:
https://github.com/aquasecurity/trivy/blob/72ea4b0632308bd6150aaf2f1549a3f10b60dc23/pkg/sbom/spdx/unmarshal.go#L186-L196

You need to add testcase (as in https://github.com/aquasecurity/trivy/pull/9126/files#diff-b17fc40bdfbe923285c5951b6ff7243b43dddedfa22eb0a5967d695268096a87).
This test case will help you to update logic.

DicksenT commented:
hello @knqyf263 @DmitriyLewen, im currently trying to update unmarshal, but quite confused on what to do.
In CycloneDX, there's a helper function (`unmarshalHashes`) that adds the digests to a field on `core.File`. But SPDX doesn't seem to have something similar right now, and I’m not sure whether I should:

1. Add a new `Digests` field to `core.File`, or if that's outside the scope.
2. Support just SHA-512 (like in the test data), or also prepare for other hashes like SHA-256 or MD5.
3. Create a separate helper function like CycloneDX, or just add the logic directly inside the `parsePackage` function.

Sorry if I missed anything obvious — would appreciate a bit of direction here

DmitriyLewen commented:
you need:
- add function to convert spdx.Checksum into digest.digest
- save checksums from `package` or `Files`
- add tests for both cases

something like that:
```diff
diff --git a/pkg/sbom/spdx/testdata/happy/package-hashes.json b/pkg/sbom/spdx/testdata/happy/package-hashes.json
index e69de29bb..6c973c8f1 100644
--- a/pkg/sbom/spdx/testdata/happy/package-hashes.json
+++ b/pkg/sbom/spdx/testdata/happy/package-hashes.json
@@ -0,0 +1,26 @@
+{
+  "SPDXID": "SPDXRef-DOCUMENT",
+  "creationInfo": {
+    "created": "2022-09-12T17:02:46.826609Z",
+    "creators": [
+      "Tool: trivy-dev",
+      "Organization: aquasecurity"
+    ]
+  },
+  "dataLicense": "CC0-1.0",
+  "documentNamespace": "http://trivy.dev/container/meven-test-project-eb7a0384-b04a-4fc6-8afb-1662fe59ca79",
+  "name": "maven-test-projecct",
+  "packages": [
+    {
+      "name": "lodash",
+      "SPDXID": "SPDXRef-Package-lodash-4.17.21",
+      "checksums": [
+        {
+          "algorithm": "SHA512",
+          "checksumValue": "bf690311ee7b95e713ba568322e3533f2dd1cb880b189e99d4edef13592b81764daec43e2c54c61d5c558dc5cfb35ecb85b65519e74026ff17675b6f8f916f4a"
+        }
+      ]
+    }
+  ],
+  "spdxVersion": "SPDX-2.3"
+}
\ No newline at end of file
diff --git a/pkg/sbom/spdx/unmarshal.go b/pkg/sbom/spdx/unmarshal.go
index 1766bdea3..b45fb0e7e 100644
--- a/pkg/sbom/spdx/unmarshal.go
+++ b/pkg/sbom/spdx/unmarshal.go
@@ -6,6 +6,8 @@ import (
 	"io"
 	"strings"
 
+	"github.com/aquasecurity/trivy/pkg/digest"
+	"github.com/aquasecurity/trivy/pkg/log"
 	"github.com/package-url/packageurl-go"
 	"github.com/samber/lo"
 	"github.com/spdx/tools-golang/json"
@@ -190,11 +192,25 @@ func (s *SPDX) parsePackage(spdxPkg spdx.Package) (*core.Component, error) {
 			{Path: path},
 		}
 	} else if len(spdxPkg.Files) > 0 {
+		file := spdxPkg.Files[0]
 		component.Files = []core.File{
-			{Path: spdxPkg.Files[0].FileName}, // Take the first file name
+			{
+				Path:    file.FileName, // Take the first file name
+				Digests: s.unmarshalChecksums(file.Checksums),
+			},
 		}
 	}
 
+	if len(component.Files) == 0 && len(spdxPkg.PackageChecksums) > 0 {
+		// If no files are specified, use the package checksums
+		component.Files = []core.File{
+			{
+				Digests: s.unmarshalChecksums(spdxPkg.PackageChecksums),
+			},
+		}
+
+	}
+
 	// Trivy stores properties in Annotations
 	// But previous versions stored properties in AttributionTexts
 	// So we need to check both cases to maintain backward compatibility
@@ -275,6 +291,27 @@ func (s *SPDX) parseExternalReferences(refs []*spdx.PackageExternalReference) (*
 	return nil, nil
 }
 
+func (s *SPDX) unmarshalChecksums(checksums []spdx.Checksum) []digest.Digest {
+	var digests []digest.Digest
+	for _, h := range checksums {
+		var alg digest.Algorithm
+		switch h.Algorithm {
+		case common.SHA1:
+			alg = digest.SHA1
+		case common.SHA256:
+			alg = digest.SHA256
+		case common.SHA512:
+			alg = digest.SHA512
+		case common.MD5:
+			alg = digest.MD5
+		default:
+			log.Warn("Unsupported hash algorithm", log.String("algorithm", string(h.Algorithm)))
+		}
+		digests = append(digests, digest.NewDigestFromString(alg, h.Value))
+	}
+	return digests
+}
+
 func (s *SPDX) isTrivySBOM(spdxDocument *spdx.Document) bool {
 	if spdxDocument == nil || spdxDocument.CreationInfo == nil || spdxDocument.CreationInfo.Creators == nil {
 		return false
diff --git a/pkg/sbom/spdx/unmarshal_test.go b/pkg/sbom/spdx/unmarshal_test.go
index 50bfcb111..ad3c26ec6 100644
--- a/pkg/sbom/spdx/unmarshal_test.go
+++ b/pkg/sbom/spdx/unmarshal_test.go
@@ -331,6 +331,18 @@ func TestUnmarshaler_Unmarshal(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:      "happy path package with hash",
+			inputFile: "testdata/happy/package-hashes.json",
+			want: types.SBOM{
+				Metadata: types.Metadata{
+					OS: &ftypes.OS{
+						Family: "alpine",
+						Name:   "3.16.0",
+					},
+				},
+			},
+		},
 		{
 			name:      "happy path empty component",
 			inputFile: "testdata/happy/empty-bom.json",

```

DicksenT committed:
Merge branch 'feat/add-spdx-sha512' of https://github.com/DicksenT/trivy into feat/add-spdx-sha512

DmitriyLewen committed:
test(marshaler): remove unneeded package from testcase

DmitriyLewen committed:
feat(spdx): save digests:
- add IsEmpty for core.File
- use pkgFiles instead of pkgFilePaths
- update tests

DmitriyLewen committed:
refactor(core): always save digest

DmitriyLewen committed:
refactor: fix default logic for converting digests

DmitriyLewen commented on pkg/sbom/io/decode.go:
```diff
@@ -235,7 +235,7 @@ func (m *Decoder) decodePackage(ctx context.Context, c *core.Component) (*ftypes
 		}
 		// An empty path represents a package digest
-		if f.Path == "" && len(f.Digests) > 0 {
+		if len(f.Digests) > 0 {
```
@knqyf263 Can we discuss about this again? :) (cf. https://github.com/aquasecurity/trivy/pull/6310#discussion_r1527746700)

afdesk replied:
@knqyf263 could you pls take a look at this comment again when you have time?
we need to close this ticket. thanks!

DmitriyLewen commented:
Hello @DicksenT 
I updated this PR.

DicksenT commented:
is there anything else i should do? @DmitriyLewen

knqyf263 reviewed (approved):
The broader question about package-level vs file-level digest handling raised on decode.go deserves its own design discussion rather than expanding this PR's scope, so I will open a GitHub Discussion to track it.

Thanks.

knqyf263 commented:
Hi @DicksenT, I approved this PR, but it was closed a few seconds later. Was that intentional? Two options, whichever works for you:

1. If the close wasn't intentional, can we reopen and merge it?
2. Otherwise, we can open a separate PR that includes your commits (with credit to you) to get it merged.

Let us know what you prefer.

DicksenT commented:
hi @knqyf263, the close was not intentional, i'll be happy if you want to reopen and merge it, however it still up to you to decide it which option is the best

knqyf263 commented:
Hi @DicksenT, thanks for confirming. I tried to reopen the PR, but GitHub doesn't let me reopen this PR.

Would you mind opening a new PR with these commits? That keeps the work under your name in the contributors list. If it's any trouble, we can open it on our side instead. Just let us know which you prefer.

Apologies for the slow turnaround on this.

DicksenT commented:
hi @knqyf263, i tried to reopen pr on here https://github.com/fossabot/trivy/pull/4/commits, but im unsure if this is correct, please check it first

no worries for the slow turnaround, cheers!

DmitriyLewen commented:
Hi @DicksenT ,
Due to the [incident and the repository rename](https://web.archive.org/web/20260306133317/https://github.com/aquasecurity/trivy/discussions/10265), all forks lost their connection to the aquasecurity repository.

We resolved this as follows:
- pull all the necessary fork data to your local PC
- delete the fork
- create the fork again (from https://github.com/aquasecurity/trivy)
- push the necessary data (in your case the branch with changes + open a PR)

We apologize for the inconvenience...
Let us know if we can help with this.

DicksenT commented:
hi @DmitriyLewen , i open a new pr #10719, please kindly check it
